### PR TITLE
fix(api): ensure module polling waits for full poll to notify

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR makes changes to how the HW API polls modules and updates its own state tracking of the hardware itself to remove concurrency bugs, increase responsiveness, and remove unnecessary code complexity. Resolves RET-1259

### Items investigated during development

- [x] Improve "disconnect on error" behavior of TC HW API
    - [x] Investigate why we disconnect driver and poller on error in existing code. Is it even necessary? (@pmoegenburg)
        - **Conclusion:** necessary for GEN1, not necessary for GEN2 (RET-1263)
- [x] (Maybe) rewrite some of the HW API unit tests to be strictly isolated unit tests rather than quasi-integration tests with partial mocking
    - **Conclusion:** not worth the effort, save this for a complete re-write
- [x] Validate that set and read is guaranteed to return the just-set value at a firmware level (@pmoegenburg)
    - [x] Heater-Shaker
    - [x] Temperature module
    - [x] Thermocycler gen1
    - [x] Thermocycler gen2
        - **Conclusion:** firmware behaves as expected: values are set immediately

## Changelog

- Calling `wait_next_poll` will now wait for the next _complete_ poll to finish
    - If `wait_next_poll` is called with a `read` is ongoing, it will be subscribed to the _next_ call to `read` rather than the current call 
- All module setters refactored to be more responsive and remove unnecessary complexity
    - Before: "set a target value and wait up to 1 second for the next poll to complete to update HW API state"
    - After: "set a target value and immediately read that value to update HW state"

## Review requests

We've been smoke testing this heavily on real hardware. The focus is on making sure all module reads and writes behave as expected. Pay special attention to any potential regression of #6392.

See ticket for a testing protocol written in Jupyter Notebook.

## Risk assessment

Medium, mitigated by:

- Thorough on-hardware testing
- The fact that this PR fixes removes code and complexity to fix the issue rather than adding code

**This PR will change the G-code that gets sent to modules.** Specifically, while write calls will remain unchanged, there will be slightly more read calls because, in addition to the poller, the HW API will now explicitly request at least one read per write

### CODE REVIEW

**Comment by mcous** [comment]
Current dependencies on/for this PR:
* edge
  * **PR #11548** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11548" target="_blank">Graphite</a>  👈
    * **PR #11573** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11573" target="_blank">Graphite</a> 
    * **PR #11550** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/11550" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/11548?utm_source=stack-comment).

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 1]
```diff
@@ -418,82 +348,71 @@ async def set_target_lid_temperature(self, celsius: float) -> None:
...
         """
         This method only exits if set temperature has been reached.
 
         Subject to change without a version bump.
         """
-        while self._listener.plate_status != TemperatureStatus.HOLDING:
+        await self._reader.read_block_temperature()
+
+        # if self.is_simulated:
+        #     return
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/thermocycler.py]: Oops, remove

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 2]
```diff
@@ -418,82 +348,71 @@ async def set_target_lid_temperature(self, celsius: float) -> None:
...
+            self._enter_error_state()
 
     @property
     def lid_target(self) -> Optional[float]:
-        return (
-            self._listener.state.lid_temperature.target
-            if self._listener.state
-            else None
-        )
+        return self._reader.lid_temperature.target if self._reader else None
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/thermocycler.py]:
```suggestion
        return self._reader.lid_temperature.target
```

File: api/src/opentrons/hardware_control/poller.py [hunk 3]
```diff
@@ -1,155 +1,74 @@
...
-        Returns: None
-        """
-        ...
+class Poller:
+    """A poller to call a given reader on an interval.
 
+    Args:
+        reader: An interface to read data.
+        interval: The poll interval, in seconds.
+        on_error: An optional callback to trigger if a poll raises an error.
```
Comment (mcous) [inline api/src/opentrons/hardware_control/poller.py]:
```suggestion
        reader: An interface to read data.
        interval: The poll interval, in seconds.
```

File: api/src/opentrons/hardware_control/modules/heater_shaker.py [hunk 4]
```diff
@@ -289,15 +288,18 @@ async def await_temperature(self, awaiting_temperature: float) -> None:
...
 
         await self.wait_for_is_running()
-        await self.wait_next_poll()
+        await self._reader.read_temperature()
 
         async def _await_temperature() -> None:
-            if self.temperature_status == TemperatureStatus.HEATING:
+            if awaiting_temperature is None:
+                while self.temperature_status != TemperatureStatus.HOLDING:
+                    await self._poller.wait_next_poll()
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/heater_shaker.py]: This is an accidental carryover from something we removed. Remove it again!

File: api/src/opentrons/hardware_control/modules/heater_shaker.py [hunk 5]
```diff
@@ -267,20 +271,15 @@ async def start_set_temperature(self, celsius: float) -> None:
...
+        await self._reader.read_temperature()
 
     # TODO(mc, 2022-10-10): remove `awaiting_temperature` argument,
     # and instead, wait until status is holding
     async def await_temperature(self, awaiting_temperature: float) -> None:
-        """Await temperature in degrees Celsius.
+        """Await temperature in degreess Celsius.
 
-        Polls temperature module's current temperature until
+        Polls the Temperature Module's current temperature until
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/heater_shaker.py]:
```suggestion
        Polls the Heater-Shaker's current temperature until
```

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 6]
```diff
@@ -647,79 +548,64 @@ def _enter_error_state(self, cause: Exception) -> None:
...
-class PolledData:
-    lid_status: ThermocyclerLidStatus
-    lid_temperature: Temperature
-    plate_temperature: PlateTemperature
+class ThermocyclerReader(Reader):
+    """Read data from the thermocycler.
 
+    Args:
+        driver: A connected Thermocycler driver.
+        handle_error: A callback to call if the
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/thermocycler.py]:
```suggestion
```

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 7]
```diff
@@ -418,82 +348,65 @@ async def set_target_lid_temperature(self, celsius: float) -> None:
...
         This method only exits if lid target temperature has been reached.
 
         Subject to change without a version bump.
         """
-        if self.is_simulated:
-            return
+        await self._reader.read_lid_temperature()
 
-        while self._listener.lid_status != TemperatureStatus.HOLDING:
+        while self._reader.lid_temperature_status != TemperatureStatus.HOLDING:
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/thermocycler.py]: Existing logic, but probably needs to account for error status to avoid a potential deadlock.

Maybe it could check that `self._reader.block_temperature is not None and self._reader.error is not None`, or something similar

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 8]
```diff
@@ -418,82 +348,65 @@ async def set_target_lid_temperature(self, celsius: float) -> None:
...
+    async def _wait_for_block_target(self) -> None:
         """
         This method only exits if set temperature has been reached.
 
         Subject to change without a version bump.
         """
-        while self._listener.plate_status != TemperatureStatus.HOLDING:
+        await self._reader.read_block_temperature()
+
+        while self._reader.block_temperature_status != TemperatureStatus.HOLDING:
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/thermocycler.py]: Same comment; seems like a deadlock risk

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 9]
```diff
@@ -264,36 +246,19 @@ async def _set_temperature_no_pause(
...
         hold_time = total_seconds if total_seconds > 0 else 0
+
         if ramp_rate is not None:
             await self._driver.set_ramp_rate(ramp_rate=ramp_rate)
+
         await self._driver.set_plate_temperature(
             temp=temperature, hold_time=hold_time, volume=volume
         )
 
-        # Wait for target temperature to be set.
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/thermocycler.py:273]: This particular block seems to have a long and storied history in our hardware control layer, but I don't see why it exists.

If you set the thermocycler's temperature, the firmware updates its state immediately and responds as such. You can confirm this by issuing the following G-code, which sets the plate temperature (`M104`) to `42` and immediately reads it back (`M105`):

```
M104 S42 M105
```

The `M105` will report the target temperature of `42.0`.

So, by switching the HW API's state update system from an "only polling" model to a "write then read back" model (while allowing the poller to continue in the background), the need for all these things (among others) in the TC HW layer seems to go away entirely:

- `self.hold_time_probably_set`
- `TEMP_UPDATE_RETRIES`
- `self._hold_time_fuzzy_seconds`

I could be wrong here, so I would love to hear someone with more history in this codebase sound off on these changes

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 10]
```diff
@@ -264,36 +246,19 @@ async def _set_temperature_no_pause(
...
-            task = self._loop.create_task(self._wait_for_hold(hold_time))
-        else:
-            task = self._loop.create_task(self._wait_for_temp())
+        task = self._loop.create_task(self._wait_for_block_target())
         self.make_cancellable(task)
         await task
 
-    async def wait_for_block_temperature(
-        self,
-        temperature: float,
```
Comment (mcous) [inline api/src/opentrons/hardware_control/modules/thermocycler.py:295]: This temperature argument was used solely to compare against the hardware's reported target. The actual wait loop waits for a status of `HOLDING`. I removed it as unnecessary for the same reason as the `hold_time_probably_set` removal

File: api/src/opentrons/hardware_control/modules/heater_shaker.py [hunk 11]
```diff
@@ -267,20 +271,15 @@ async def start_set_temperature(self, celsius: float) -> None:
...
-        # Re-think this pattern, potentially even at the driver/firmware level
         await self._driver.set_temperature(celsius)
-        await self.wait_next_poll()
+        await self._reader.read_temperature()
 
     # TODO(mc, 2022-10-10): remove `awaiting_temperature` argument,
     # and instead, wait until status is holding
     async def await_temperature(self, awaiting_temperature: float) -> None:
-        """Await temperature in degrees Celsius.
+        """Await temperature in degreess Celsius.
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/heater_shaker.py]:
```suggestion
        """Await temperature in degrees Celsius.
```

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 12]
```diff
@@ -348,40 +311,19 @@ async def cycle_temperatures(
...
-        task = self._loop.create_task(self._wait_for_lid_temp())
+
+        task = self._loop.create_task(self._wait_for_lid_target())
         self.make_cancellable(task)
         await task
 
-    async def wait_for_lid_temperature(self, temperature: float) -> None:
-        """Set the lid temperature in deg Celsius"""
+    async def wait_for_lid_target(self) -> None:
+        """Set the lid temperature in degres Celsius"""
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/thermocycler.py]:
```suggestion
        """Set the lid temperature in degrees Celsius"""
```

**Review by sfoster1 — APPROVED** [review]
This looks like a much much better approach to this, yeah. Nice.

File: api/src/opentrons/drivers/thermocycler/simulator.py [hunk 13]
```diff
@@ -48,7 +48,7 @@ async def set_plate_temperature(
     ) -> None:
         self._plate_temperature.target = temp
         self._plate_temperature.current = temp
-        self._plate_temperature.hold = hold_time
```
Comment (sanni-t) [inline api/src/opentrons/drivers/thermocycler/simulator.py:51]: Why this change?

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 14]
```diff
@@ -634,9 +544,12 @@ async def _execute_cycles(
             for step_idx, step in enumerate(steps):
                 self._current_step_index = step_idx + 1  # science starts at 1
                 await self._execute_cycle_step(step, volume)
-                await self._wait_for_hold()
 
-    def _enter_error_state(self, cause: Exception) -> None:
+    # TODO(mc, 2022-10-13): why does this exist?
+    # Do the driver and poller really need to be disconnected?
+    # Could we accomplish the same thing by latching the error state
+    # and allowing the driver and poller to continue?
```
Comment (sanni-t) [inline api/src/opentrons/hardware_control/modules/thermocycler.py:551]: Replace this TODO with the answers to the listed questions?

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 15]
```diff
@@ -36,6 +35,16 @@ class ThermocyclerError(Exception):
     pass
 
 
+def _temperature_is_holding(status: Optional[TemperatureStatus]) -> bool:
+    if status in (TemperatureStatus.HOLDING, TemperatureStatus.IDLE):
```
Comment (sanni-t) [inline api/src/opentrons/hardware_control/modules/thermocycler.py:39]: Why is a status of `IDLE` being considered as 'holding temperature'? It feels like it could lead to bugs; although, with the way `_temperature_is_holding` is used, I don't think that the status arg would ever be `IDLE`. But it also doesn't seem necessary. Am I missing something?

File: api/src/opentrons/hardware_control/modules/thermocycler.py (hunk 9 above)
Comment (sanni-t) [reply api/src/opentrons/hardware_control/modules/thermocycler.py:273]: I might be forgetting some finer details but you are correct in that switching to a 'write then read back' technique would eliminate the need for all of this complicated verification. It was required precisely because we were relying on polling to update the internal hardware controller state. We'd run into situations where if the new set_temperature command sets the same target as previous one, except with a new hold time, then the vanilla comparison logic of `assert new_hold_time == target_hold_time_read_back` would fail because by the time the poller read the hold_time back, it would have already changed from the value it was set to.
tl;dr- this change looks good

**Review by sanni-t — APPROVED** [review]
Still have some tests to go through but the code changes look good to me so you can go ahead with merging if I don't get the test review in in time. Just have some nits and questions.
Also, we should test error detection in TC gen1 & H/S if you guys haven't done that already (I can help with that if needed).

File: api/src/opentrons/drivers/thermocycler/simulator.py (hunk 13 above)
Comment (mcous) [reply api/src/opentrons/drivers/thermocycler/simulator.py:51]: If the intention with the simulator is to simulate "module immediately set to hold completed state," then the reported hold time should be `0` (aka "hold finished"), not the requested hold time.

This change removes the need for at least one `if self.simulating: return` block in the TC HW API

File: api/src/opentrons/hardware_control/modules/thermocycler.py (hunk 15 above)
Comment (mcous) [reply api/src/opentrons/hardware_control/modules/thermocycler.py:39]: Yeah I agree it feels a little weird, but I want something in there for completeness, since `TemperatureStatus` can be one of five values:

```py
class TemperatureStatus(str, Enum):
    HOLDING = "holding at target"
    COOLING = "cooling"
    HEATING = "heating"
    IDLE = "idle"
    ERROR = "error"
```

I didn't want to leave `IDLE` it out and leave ourselves open to an infinite loop, so "holding" felt like the most appropriate. I could also move it to a `raise`, if that feels more correct to you

File: api/src/opentrons/hardware_control/modules/thermocycler.py (hunk 14 above)
Comment (mcous) [reply api/src/opentrons/hardware_control/modules/thermocycler.py:551]: Will address in #11550 for merge ease

File: api/src/opentrons/hardware_control/modules/thermocycler.py (hunk 15 above)
Comment (sanni-t) [reply api/src/opentrons/hardware_control/modules/thermocycler.py:39]: Ah, makes sense. Ya, I think a `raise` would help tie the loose ends.
